src/main/java/io/leonis/subra/ipc/serialization/protobuf/SSLVisionDeducer.java

```diff
@@ -0,0 +1,46 @@
+package io.leonis.subra.ipc.serialization.protobuf;
+
+import io.leonis.subra.ipc.serialization.protobuf.SSLVisionDeducer.VisionPacket;
+import io.leonis.subra.ipc.serialization.protobuf.vision.*;
+import java.util.Set;
+import lombok.*;
+import org.reactivestreams.Publisher;
+import io.leonis.subra.game.data.*;
+import io.leonis.zosma.game.engine.*;
+import io.leonis.algieba.Temporal;
+import org.robocup.ssl.Wrapper.WrapperPacket;
+import reactor.core.publisher.Flux;
+
+/**
+ * The Class SSLVisionDeducer.
+ *
+ * This class represents a {@link Deducer} which creates {@link VisionPacket} from {@link
+ * WrapperPacket}.
+ *
+ * @author Jeroen de Jong
+ * @author Rimon Oz
+ */
+@AllArgsConstructor
+public class SSLVisionDeducer implements Deducer<WrapperPacket, VisionPacket> {
+
+  @Override
+  public Publisher<VisionPacket> apply(final Publisher<WrapperPacket> wrapperPacketPublisher) {
+    return Flux.from(wrapperPacketPublisher)
+        .transform(new ParallelDeducer<>(
+            new PlayersDeducer(),
+            new GoalsDeducer(),
+            new BallsDeducer(),
+            new FieldDeducer(),
+            VisionPacket::new));
+  }
+
+  @Value
+  public static class VisionPacket
+      implements Player.SetSupplier, Goal.SetSupplier, Field.Supplier, Ball.SetSupplier, Temporal {
+    private final Set<Player> agents;
```

Comment:
your rename failed; it's still called agents here.

Reply:
it has to be called agents here, because it's a Player.SetSupplier which extends Agent.SetSupplier, which has a `getAgents()`

Reply:
kuch leaky abstraction

Reply:
we can do with not overriding Agent.SetSupplier and removing it from zosma entirely. i don't feel strongly one way or the other

Reply:
There are two problems. 
 1. A generic Agent-dependent system can't work if you dont override the getter
 1. The Player.SetSupplier is an actual dependency since you usually need a specific type of agent. 


Do we use the Agent.SetSupplier somewhere reasonable currently?

Reply:
we don't use it anywhere except for composing Player.SetSupplier

Reply:
Should we remove it?